### ISSUE TITLE
Invalid refs should throw

### ISSUE THREAD

**Issue by inverted-capital** [isomorphic-git/isomorphic-git#1885]
In this code, provision is made for new branches with no commits by assigning the special null tree hash: https://github.com/isomorphic-git/isomorphic-git/blob/fb8a29a32529fef8dcf6c3865e2a304510ec88be/src/models/GitWalkerRepo.js#L23

This does not cover the case where a completely invalid ref has been passed in, and worse, this fails silently.

I am unsure how to correct this, but I think an error should be thrown, or at least a warning emitted ?

**Comment by jcubic** [comment]
I'm not sure what to say. Do you have a real case where this is a problem?

**Comment by inverted-capital** [comment]
I don't have a repro yet, but I will make one soon.  The exact issue I had was that I was sending in a ref as a string with a newline at the end.  The check for valid sha simply checks if the length is 40 characters, which failed, and then the special oid representing an empty tree `4b825...` was used instead, which caused a silent fail in my case.

Where is the reference material that you guys use to define behaviours for what isomorphic-git should do in response to given parameters ?  Basically, where should I go to discern what the correct behaviour in this instance is ?

**Comment by jcubic** [comment]
for your usecase  this can be fixed by updating this regex:

https://github.com/isomorphic-git/isomorphic-git/blob/main/src/managers/GitRefManager.js#L47-L51

It should be `/^[0-9a-f]{40}$/`

**Comment by inverted-capital** [comment]
Good spotting.  But what is the correct behavior for when the ref is for a new branch with no commits ?  Where do we get the canonical behavior from ?  the git.exe file that is published by https://git-scm.com/ ?

Now that I go looking, there doesn't appear to be an actual spec for git 🤷

**Comment by jcubic** [comment]
I wonder what will happen when `await resolveTree` get this fake oid. Maybe this is not needed at all.

This requires some investigation, what happen if you run the code in empty repo and on invalid ref.

I think that the best way to get canonical behavior, is to just use the canonical git and test. I personally use Linux and git came from the official repository.

**Comment by inverted-capital** [comment]
Ok thanks.  For reference, the oid `4b825dc642cb6eb9a060e54bf8d69288fbee4904` is special, in that it is the oid of an empty tree, just in case that wasn't clear.

**Comment by jcubic** [comment]
But is it part of the git spec?

**Comment by inverted-capital** [comment]
yes, 100% part of the reference git implementation as this is the result of doing a sha1 hash on an empty tree object.

**Comment by inverted-capital** [comment]
Thank you - I will get to it when I can.
